engine/src/test/java/io/camunda/zeebe/engine/state/migration/VersionCompatibilityCheckTest.java

```diff
@@ -0,0 +1,99 @@
+import static org.assertj.core.api.Assertions.assertThat;
+
+import io.camunda.zeebe.engine.state.migration.VersionCompatibilityCheck.CheckResult.Compatible;
+import io.camunda.zeebe.engine.state.migration.VersionCompatibilityCheck.CheckResult.Incompatible;
+import io.camunda.zeebe.engine.state.migration.VersionCompatibilityCheck.CheckResult.Indeterminate;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.parallel.Execution;
+import org.junit.jupiter.api.parallel.ExecutionMode;
+
+@Execution(ExecutionMode.CONCURRENT)
+final class VersionCompatibilityCheckTest {
+
+  @Nested
+  final class IncompatibleResults {
+
+    @Test
+    public void shouldRejectDowngrades() {
+      assertThat(check("9.0.0", "8.0.0")).isInstanceOf(Incompatible.MajorDowngrade.class);
+      assertThat(check("9.0.0", "8.1.0")).isInstanceOf(Incompatible.MajorDowngrade.class);
+      assertThat(check("8.5.0", "8.4.0")).isInstanceOf(Incompatible.MinorDowngrade.class);
+      assertThat(check("8.5.0", "8.4.1")).isInstanceOf(Incompatible.MinorDowngrade.class);
+      assertThat(check("8.5.3", "8.5.2")).isInstanceOf(Incompatible.PatchDowngrade.class);
+      assertThat(check("8.5.3", "8.5.0")).isInstanceOf(Incompatible.PatchDowngrade.class);
+      assertThat(check("8.5.3-alpha2", "8.5.3-alpha1"))
+          .isInstanceOf(Incompatible.PreReleaseDowngrade.class);
+    }
+
+    @Test
+    public void shouldRejectMajorUpgrades() {
+      assertThat(check("8.0.0", "9.0.0")).isInstanceOf(Incompatible.MajorUpgrade.class);
+      assertThat(check("8.1.0", "9.0.0")).isInstanceOf(Incompatible.MajorUpgrade.class);
+    }
+
+    @Test
+    public void shouldRejectSkippedMinorVersions() {
+      assertThat(check("8.0.0", "8.2.0")).isInstanceOf(Incompatible.SkippedMinorVersion.class);
+      assertThat(check("8.0.2", "8.3.0")).isInstanceOf(Incompatible.SkippedMinorVersion.class);
+    }
+  }
+
+  @Nested
+  final class IndeterminateResults {
+    @Test
+    void shouldDetectUnknownVersions() {
+      assertThat(check(null, "8.0.0")).isInstanceOf(Indeterminate.PreviousVersionUnknown.class);
+      assertThat(check("8.0.0", null)).isInstanceOf(Indeterminate.CurrentVersionUnknown.class);
+    }
+
+    @Test
+    void shouldDetectInvalidVersions() {
+      assertThat(check("dev", "1.2.3")).isInstanceOf(Indeterminate.PreviousVersionInvalid.class);
+      assertThat(check("1.2.3", "dev")).isInstanceOf(Indeterminate.CurrentVersionInvalid.class);
+    }
+  }
+
+  @Nested
+  final class CompatibleResults {
+    @Test
+    void shouldAcceptPreReleaseUpgrades() {
+      assertThat(check("8.0.0-alpha1", "8.0.0-alpha2"))
+          .isInstanceOf(Compatible.PreReleaseUpgrade.class);
+      assertThat(check("8.0.0-alpha1", "8.0.0-alpha4"))
+          .isInstanceOf(Compatible.PreReleaseUpgrade.class);
+      assertThat(check("8.0.0-alpha1", "8.0.0-beta1"))
+          .isInstanceOf(Compatible.PreReleaseUpgrade.class);
+    }
```

Comment:
🤔 Interesting. Currently, we don't support upgrading pre-release versions, right? Not sure whether we want to allow it here.

❌ Please add a test case that shows that you cannot upgrade from a pre-release to a regular release version, e.g. reject upgrade for `8.0.0-alpha3 -> 8.0.0`

Reply:
That's a good point that I'm not sure about :+1: 

I think it's a grey area currently because we don't guarantee that this works but probably make use of it sometimes in internal testing etc.

Reply:
Implemented as discussed in https://github.com/camunda/zeebe/pull/16421/commits/15e68ded2da3963580fa7cbb2fdb1a420b8725da